external/web3-blockchain-service.js

```diff
@@ -11,10 +11,48 @@ class Web3BlockchainService {
         this.gasStationLink = 'https://gasstation-mumbai.matic.today/v2';
     }
 
-    initialize(logger) {
+    async initialize(logger) {
         this.logger = logger;
         this.rpcNumber = 0;
-        this.web3 = new Web3(this.config.rpcEndpoints[0]);
+        await this.initializeWeb3();
+        await this.initializeContracts();
+    }
+
+    async initializeWeb3() {
+        let tries = 0;
+        let isRpcConnected = false;
+        while (!isRpcConnected) {
+            if (tries >= this.config.rpcEndpoints.length) {
+                throw Error('Blockchain initialisation failed');
+            }
+
+            try {
+                this.web3 = new Web3(this.config.rpcEndpoints[this.rpcNumber]);
+                isRpcConnected = await this.web3.eth.net.isListening();
+            } catch (e) {
+                this.logger.warn(
+                    `Unable to connect to blockchain rpc : ${
+                        this.config.rpcEndpoints[this.rpcNumber]
+                    }.`,
+                );
+                tries += 1;
+                this.rpcNumber = (this.rpcNumber + 1) % this.config.rpcEndpoints.length;
```

Comment:
If i understand correctly we have fallback mechanism only for the web3 initialization. We should also add fallback mechanism when calling callContractFunction and executeContractFunction, basically whenever we are using web3 object

Reply:
The fallback is also on execution of smart contract functions. handleError() will restart the service if the rpc is not connected

Reply:
https://github.com/OriginTrail/ot-node/blob/7fd575f2d7efae47b1edfef2dcb59e71b66cd157/external/web3-blockchain-service.js#L195-L209